zepid/causal/doublyrobust/TMLE.py

```diff
@@ -800,15 +801,537 @@ def plot_love(self, color_unweighted='r', color_weighted='b', shape_unweighted='
+
+    .. math::
+
+        E[\text{logit}(Q(A=a, W)) + \hat{\epsilon}]
+
+    This process is repeated a large number of times and the point estimate is the average of those individual treatment
+    plans.
+
+    Examples
+    --------
+    Setting up environment
+
+    >>> from zepid import load_sample_data, spline
+    >>> from zepid.causal.doublyrobust import StochasticTMLE
+    >>> df = load_sample_data(False).dropna()
+    >>> df[['cd4_rs1', 'cd4_rs2']] = spline(df, 'cd40', n_knots=3, term=2, restricted=True)
+
+    Estimating TMLE for 0.2 being treated with ART
+
+    >>> tmle = StochasticTMLE(df, exposure='art', outcome='dead')
+    >>> tmle.exposure_model('male + age0 + cd40 + cd4_rs1 + cd4_rs2 + dvl0')
+    >>> tmle.outcome_model('art + male + age0 + cd40 + cd4_rs1 + cd4_rs2 + dvl0')
+    >>> tmle.fit(p=0.2)
+    >>> tmle.summary()
+
+    Estimating TMLE for conditional plan
+
+    >>> tmle = StochasticTMLE(df, exposure='art', outcome='dead')
+    >>> tmle.exposure_model('male + age0 + cd40 + cd4_rs1 + cd4_rs2 + dvl0')
+    >>> tmle.outcome_model('art + male + age0 + cd40 + cd4_rs1 + cd4_rs2 + dvl0')
+    >>> tmle.fit(p=[0.6, 0.4], conditional=["df['male']==1", "df['male']==0"])
+    >>> tmle.summary()
+
+    Estimating TMLE with machine learning algorithm from sklearn
+
+    >>> from sklearn.linear_model import LogisticRegression
+    >>> log1 = LogisticRegression(penalty='l1', random_state=201)
+    >>> tmle = StochasticTMLE(df, 'art', 'dead')
+    >>> tmle.exposure_model('male + age0 + cd40 + cd4_rs1 + cd4_rs2 + dvl0', custom_model=log1)
+    >>> tmle.outcome_model('male + age0 + cd40 + cd4_rs1 + cd4_rs2 + dvl0', custom_model=log1)
+    >>> tmle.fit(p=0.75)
+
+    References
+    ----------
+    Muñoz ID, and Van Der Laan MJ. Population intervention causal effects based on stochastic interventions.
+    Biometrics 68.2 (2012): 541-549.
+
+    van der Laan MJ, and Sherri R. Targeted learning in data science: causal inference for complex longitudinal
+    studies. Springer Science & Business Media, 2011.
+    """
+    def __init__(self, df, exposure, outcome, alpha=0.05, continuous_bound=0.0005, verbose=False):
+        # Dropping ALL missing data (currently doesn't allow for censored outcomes)
+        if df.dropna().shape[0] != df.shape[0]:
+            warnings.warn("There is missing data in the data set. StochasticTMLE will drop all missing data. "
+                          "StochasticTMLE will fit "
+                          + str(df.dropna().shape[0]) +
+                          ' of ' + str(df.shape[0]) + ' observations', UserWarning)
+            self.df = df.copy().dropna().reset_index()
+        else:
+            self.df = df.copy().reset_index()
+
+        if not df[exposure].value_counts().index.isin([0, 1]).all():
+            raise ValueError("StochasticTMLE only supports binary exposures currently")
+
+        # Manage outcomes
+        if df[outcome].dropna().value_counts().index.isin([0, 1]).all():
```

Comment:
how often do you do this "trick"? I feel like probably this is repeated many times in zepid? Maybe time to wrap into a function?

Reply:
often enough that it would be put into a function